### DISCUSSION TITLE
packaging/deltas

### DISCUSSION THREAD

**Discussion by giscus** [velopack/velopack.docs#26] [General]
# packaging/deltas

Delta updates provide speedy updates for even the biggest of applications.

https://docs.velopack.io/packaging/deltas

**Comment by wozaiha** [comment]
Do I need to keep those old `xxx-full.nupkg` if I want a delta update?
Or just those delta files is enough?

**Reply by caesay** [reply]
In general I recommend keeping full/delta releases in pairs, and deleting them together at the same time. While there's nothing to prohibit updates working with just the delta files, it's not a tested/supported scenario and could break at some point.

**Comment by cesar-albusac** [comment]
I am getting an strange behavior: the program size (in Programs & Files) on windows increases after an update, even though I didn't add any changes to the Release. 

1. I created my first Release, 1.0.0 and install it on my PC. 
    Windows MyApp Size: 537 MB
    Windows LocalAppData folder size: 537 MB

2. I created a second Release (no code changes) 2.0.0 and start my app 1.0.0 so I get an upgrade to 2.0.0. 
    Windows MyApp  Size: 910 MB
    Windows LocalAppData folder size: 537 MB

**Reply by caesay** [reply]
The documentation website comments is not the place to get support on specific issues (I do not have enough information to diagnose your problem, nor will the answer be very useful to others). In general, your application size could increase a little after the first update due to how Velopack manages local packages and deltas. Check your app directory for any files you don't expect, and if you have any further questions please open a GitHub Issue, or join our Discord.

**Comment by cesar-albusac** [comment]
What's the best way to keep multiple different stages of an installer? 

Let's say I have MyApp v1 (branch Release_1), which started on 1.0.0 , v2 (branch Release_2) with 2.0.0 , and current which is 3.0.0 (branch Release). 

I guess we want to keep the Releases and packages output separated so delta al calculated properly? 

Otherwise, if we are creating new versions of v1 , v2 and v3 at the same time, it won't know how to calculate the delta packages from the previous ones. 

Any thoughts?

**Reply by caesay** [reply]
You should use [channels](https://docs.velopack.io/packaging/channels) for this. You don't need to separate anything further than just building each of those on a different channel.

**Comment by svstoichkov** [comment]
If only part of a file changes between versions, does Velopack download the entire updated file or just a delta containing the changes to that specific file? Let's say I have a 1GB binary file and 10KB of it changes, will Velopack only download 10KB and apply them over the existing file or will it download the whole 1GB?

**Reply by caesay** [reply]
It will only download a diff (the 10kb of the file that changed). The one limitation is our current diff algorithm is limited to a maximum of 2GB per single file. If you have files larger than that, deltas will be disabled/not be possible. I hope to improve this in the future.

**Reply by svstoichkov** [reply]
Thank you. If I have a 3GB file will deltas be disabled only for that file or they will be disabled for the whole update package in general?

**Reply by caesay** [reply]
Right now, if you have a single file larger than 2GB deltas will be disabled for the whole update package. The total size of the package may be bigger than 2GB, just not any one file.
